hail/python/hail/matrixtable.py

```diff
@@ -2490,7 +2491,15 @@ def write(self, output: str, overwrite: bool = False, stage_locally: bool = Fals
             If ``True``, overwrite an existing file at the destination.
         """
 
-        writer = MatrixNativeWriter(output, overwrite, stage_locally, _codec_spec)
+        if _partitions:
+            partitions_type = hl.tarray(hl.tinterval(self.row_key.dtype))
+            if isinstance(_partitions, Expression):
+                if _partitions.dtype != partitions_type:
+                    raise ValueError(f'_partitions has wrong type: got {_partitions.dtype}, expected {partitions_type}')
```

Comment:
This again feels like too strict a check. If the point type of the interval is a prefix of the key, or a value type that could be wrapped in a struct to be a prefix of a key, we should allow it. That's certainly what `RVDPartitioner`'s constructor does.